src/pinnwand/defensive.py

```diff
@@ -50,6 +54,31 @@ def ratelimit(request: HTTPServerRequest, area: str = "global") -> bool:
     return False
 
 
+def ratelimit(area: str):
+    """A ratelimiting decorator for tornado's request handlers."""
+
+    def wrapper(func):
+        @wraps(func)
+        def inner(*args, **kwargs):
+            # We transform the args into a dict because ChainMap works differently starting python 3.9
+            # In previous versions, it expects all maps to be dictionaries, whereas starting py3.9
+            # It transforms list/tuples into a dict whose keys are the different elements.
+            arguments = ChainMap(dict.fromkeys(args), kwargs.values())
+            request_handler = None
+            for arg in arguments:
+                if isinstance(arg, RequestHandler):
+                    request_handler = arg
+                    break
+
+            if should_be_ratelimited(request_handler.request, area):
+                raise error.RatelimitError()
+            return func(*args, **kwargs)
```

Comment:
Since it's guaranteed that the first argument will always be self which is the request handler, does this work?
```suggestion
        @wraps(func)
        def inner(self, *args, **kwargs):
            if should_be_ratelimited(self.request, area):
                raise error.RatelimitError()
                
            return func(self, *args, **kwargs)
```

Reply:
I first had it like that, by precising that the first argument is instance of RequestHandler, but then some tests failed because it got other unexpected arguments, so i just i thought I'd make it generic.

I prefer the explicitness of the parameter as well, so a combination the one you're suggesting should work.

I'll try that out

Reply:
Fixed in the forcepush 002b21f97a92d645daadc91966707764e1310874